File: csharp/1620.coordinate-with-maximum-network-quality.cs
/*
 * @lc app=leetcode id=1620 lang=csharp
 *
 * [1620] Coordinate With Maximum Network Quality
 */

using System;
using System.Linq;

// @lc code=start
public class Solution {
    public int[] BestCoordinate(int[][] towers, int radius) {
        int n = towers.Length, q = 0;
        int[] ret = new int[2];
        for (int i = 0; i < 51; ++i) {
            for (int j = 0; j < 51; ++j) {
                int acc = 0;
                for (int k = 0; k < n; ++k) {
                    double d = Math.Sqrt(Math.Pow((double)i - (double)towers[k][0], 2) + Math.Pow((double)j - (double)towers[k][1], 2));
                    if (d > (double)radius) {
                        continue;
                    }
                    acc += (int)(towers[k][2] / (1 + d));
                }
                if (acc > q) {
                    q = acc;
                    ret[0] = i;
                    ret[1] = j;
                }
            }
        }
        return ret;
    }
}
// @lc code=end

File: java/1620.coordinate-with-maximum-network-quality.java
/*
 * @lc app=leetcode id=1620 lang=java
 *
 * [1620] Coordinate With Maximum Network Quality
 */

// @lc code=start
class Solution {
    public int[] bestCoordinate(int[][] towers, int radius) {
        int n = towers.length, retX = -1, retY = -1;
        long max = -1;
        for (int x = 0; x <= 50; ++x) {
            for (int y = 0; y <= 50; ++y) {
                long sig = 0;
                for (int i = 0; i < n; ++i) {
                    double d = Math.sqrt(Math.pow((double)x - (double)towers[i][0], 2) + Math.pow((double)y - (double)towers[i][1], 2));
                    if (d > radius) {
                        continue;
                    }
                    sig += (long)Math.floor((double)towers[i][2] / (1.0 + d));
                }
                if (sig > max) {
                    max = sig;
                    retX = x;
                    retY = y;
                }
            }
        }
        return new int[] { retX, retY };
    }
}
// @lc code=end

